app/grandchallenge/discussion_forums/migrations/0002_migrate_old_forums.py

```diff
@@ -25,45 +186,76 @@ def migrate_challenge_forums(apps, schema_editor):
+        except ObjectDoesNotExist:
+            new_forum = Forum.objects.create(
+                source_object=challenge.forum,
+                created=challenge.forum.created,
+                modified=challenge.forum.updated,
             )
+            challenge.discussion_forum = new_forum
+            challenge.save()
+            init_forum_permissions(apps=apps, challenge=challenge)
+
+        topic_count = 0
+        post_count = 0
+
+        for topic in challenge.forum.topics.all():
+            try:
+                new_topic = topic.migrated_topic
+            except ObjectDoesNotExist:
+                new_topic = ForumTopic.objects.create(
+                    source_object=topic,
+                    forum=new_forum,
+                    creator=topic.poster,
+                    subject=topic.subject,
+                    kind=topic_type_matching_dict[topic.type],
+                    last_post_on=topic.last_post_on,
+                    created=topic.created,
+                    modified=topic.updated,
+                    is_locked=topic_lock_matching_dict[topic.status],
+                )
+                init_topic_permissions(apps=apps, topic=new_topic)
+                topic_count += 1
 
             latest_new_post = None
-            for post in topic.posts.all():
+            for post in topic.posts.filter(migrated_post__isnull=True).all():
                 # posts are ordered by creation time, so the last post we create here
                 # will correspond to the latest post and can be saved as that on the
                 # topic
-                latest_new_post = ForumPost.objects.create(
-                    topic=new_topic,
-                    created=post.created,
-                    modified=post.updated,
-                    creator=post.poster,
-                    content=post.content,
-                )
+                try:
+                    latest_new_post = post.migrated_post
+                except ObjectDoesNotExist:
+                    latest_new_post = ForumPost.objects.create(
+                        source_object=post,
+                        topic=new_topic,
+                        created=post.created,
+                        modified=post.updated,
+                        creator=post.poster,
+                        content=post.content,
+                    )
+                    init_post_permissions(apps=apps, post=latest_new_post)
+                    post_count += 1
+
+            if latest_new_post:
+                new_topic.last_post = latest_new_post
+                new_topic.save()
 
-            new_topic.last_post = latest_new_post
-            new_topic.save()
+        print(
+            f"Finished migrating forum, topics and posts for {challenge.short_name}. Migrated {topic_count} topics and {post_count} posts"
+        )
 
 
-def migrate_topic_tracks(apps, schema_editor):
+def migrate_topic_tracks(apps, schema_editor):  # noqa: C901
```

Comment:
Can this handle partial applications of the migration?

Reply:
What would be a partial migration in this case?